### PR TITLE
Tuples first pass

### PR DESCRIPTION
**Pull request by alexswan10k**
Tried to stick to my lane this time so hopefully I haven't added to the pile of problems!

* Fixed string concatenation
* Got tuples working incl struct tuples (they are just not wrapped)
* Got anon types working with a rather hacky "just use tuples" approach. Doesn't read great, but at least code compiles. Can revisit in the future
* Made some improvements to maybeCloneRef, turns out there was an obvious optimization we can do to remove some .clone() spam.

Output is definitely starting to look a bit less dreadful. Still frustrated by the let binding arrow problem though, not sure how to fix that!

### CODE REVIEW

**Comment by ncave** [comment]
@alexswan10k This is great, especially the `clone()` optimization!

> let binding arrow problem though, not sure how to fix that

Not sure what you mean, can you give an example?
I've already added cloning for let bindings and enabled the test that was breaking before, it should work now.

**Comment by alexswan10k** [comment]
Nice.

Yeah so if you do something like this in F#:
```
let a = something 
let b = something else
let c = something else
```

you end up with this sort of thing:
```
let a = something ;
{
   let b = something else ;
   {
      let c = something else;
      {
           ....
      }
   }
}
```

The reason it is a bit fiddly to fix is the fable AST has a let binding defined as (expr, body), so a single body is often actually a bunch of statements ending in a return. I think when I was messing around with the Lua transpiler, I did some hack to flatten it out, but it was far from elegant.

Anyway it runs fine so it is hardly a priority, it just gets rather hard to follow when you are about 7 statements deep!

**Comment by ncave** [comment]
@alexswan10k I see, so it's just about the nesting. Yes, it occasionally happens with `Sequential` too, not just `Let`.
Yes, it would be nice to flatten where possible, but not a priority as you said.

**Comment by ncave** [comment]
@alexswan10k

I flattened the `Let` and `Sequential`, hopefully it doesn't break anything. It's not ideal:
- it's not tail recursive, so it may fail on really big ASTs that only consist of `Let` 😃.
- only simplifies patterns like `Let(Let` or `Sequential(Sequential`, or `Let(Sequential`.
- it's not going to simplify `Let(sequential(Let(Sequential` patterns.
- I'm not really sure about the flattening order of the `Sequential` tree, I went with depth first.

It seems to work for now. It can always be removed if needed.

**Comment by alexswan10k** [comment]
Ah excellent, great work. I was thinking of taking on closures next by the way, you happy with that? 

Closures is going to open up Pandora's box of move problems, so I think it's a good one to nail sooner rather than later. It also opens the door to writing libs for list and array where functions are passed as args. 

I was thinking it is probably also time to build a crude reference tracker that runs on the ast before the fable2rust transform. From this you can then work out if something has moved into a closure, and thus clone and bind before throwing over the wall. The tracker would gather context about instances and track this in the compiler object. It can eventually also be used to remove more clones but that is a problem for another day. What do you think?

Not sure when il get to this next though but will try to keep it contained heh.

**Comment by ncave** [comment]
@alexswan10k

Sure, hack away! It will be nice to have a test that breaks on closures, then work on fixing it.
For AST analysis there is `FableTransforms.deepExists`, hopefully it helps.

**Comment by alexswan10k** [comment]
Ok I am a little stuck...

I need to get to this:
```
pub fn map<a_, b_, F: Fn(a_) -> b_>(f: F, x: a_) -> b_ { f(x) }
```

This is what we get today:
```
pub fn map<a_, b_>(f: fn(a: a_) -> b_, x: a_) -> b_ { f(x) }
```
from the following f#
```
let map f x =
    f x
```

the key thing is F is constrained to a trait of Fn(parameter types) -> returnType. What does this look like on the output AST? I am digging around in Rust.AST.Types.GenericBound.Trait but struggling to figure out how I would create this as I cannot see any placeholders to put in actual types, is this correct? I appreciate the output AST and printer might be missing some stuff here by the way, let me know if I need to pad anything in.

This is proving to be quite hard, I think I am going to have to rewrite the function pointer type to a generic param constrained to a trait of Fn, so that functions can accept closures also. If you try to pass a closure capturing something to a thing expecting a function pointer it blows up, as presumably it has nowhere to put the struct representing the closed over variables, just a pointer to the underlying function address. I guess we will also want this to eventually be FnMut to allow maximum flexibility?

Happy to put up code if you are curious, but it doesn't work lol.

**Comment by ncave** [comment]
@alexswan10k 
The AST types and printer should not be missing anything, they're ported from the Rust compiler.

>What does this look like on the output AST

I usually use this to see what the Rust AST looks like:
`rustc -Z unpretty=ast-tree src/main.rs > temp/ast_tree.json`
